Migrate unread message count to react

nicksellen opened the pull request:
#### Proposed Changes

* migrate the unread messages counter service and component to react

Originally it was just about the unread messages counter component, but I decided to also migrate the polling service away from angular too.

It's now a bit more separated, but potentially feels more complex as there are 3 seperate services. But they hopefully have clear roles! I'm a bit confused lately generally, so would be glad to hear from others if they find it makes sense or not, I don't know if I can tell any more :/

I made them all so they have enable/disable functions so we have more control over when they begin their operation (as opposed to starting as soon as the module is imported). At the moment the angular `run` block starts them.

The testing code was a bit annoying actually, as all the angular-compat stuff needs mocking away, and I needed `jest.isolateModules()` to probably reset the state of a service file between tests. The tests are somewhat minimal.

Also, regarding testing, to make them work with all the existing tests we basically bootstrap the core of the app automatically in `jest/jest.setup.js` but eventually it would be nice to to remove that setup and have each test loads what it needs.

The services are:

**unread mesage count service**

The job of this is to poll for unread messages counts when visible and logged in.

**visibility service**

Use the [Page Visiblity API](https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API) to check if the page is in the foreground/background for two purposes:
- lower the polling rate when not visible
- refresh the count immediately when returning to foreground

The old way used the page blur events, this replaces that.

**favicon updater service**

Just updates the favicon when the count changes.

Also, involved adding some utilities:

**dom ready**

Calls the callback when the DOM is ready! (like jQuery function). Useful if doing DOM stuff - used here for the favicon updater...

```js
import { ready } from '@/modules/core/client/utils/dom';

ready(() => {
  const el = document.getElementById('foo');
});
```

**createSubscribable**

I like the simple pattern for have subscribe/notify behaviour.

```js
import createSubscribable from '@/modules/core/client/utils/subscribable';

 const { subscribe, notify } = createSubscribable();
 
 // add a subscriber, keeping hold of the unsubscribe function
 const unsubscribe = subscribe(value => console.log('received value', value));
 
 // notify all subscribers with a value
 notify('some value');

 // unsubscribe our subscriber
 unsubscribe();
```

There is probably some library in npm that does this, but they often have loads of extra features, etc...

#### Testing Instructions

* `git switch migrate/unread-message-count`
* `npm start`
* you need two accounts in seperate browsers/sessions
* write messages to each other, leaving them unread for one of the users
* see if the little red indicator on messages changes
  ![image](https://user-images.githubusercontent.com/7449720/79145851-a0e8e880-7dc1-11ea-8115-26cab4b722bc.png)
* see if the favicon changes
  ![image](https://user-images.githubusercontent.com/7449720/79146013-ee655580-7dc1-11ea-9e4b-e81a053efc5e.png)

The polling interval is 30s for foreground, and 300s for background, so be a little patient.

Fixes #

simison commented:
I'll get back to review this but just as a thought, did you consider something like https://socket.io/ here?

A Live-time socket could be interesting in the future for real-time-messaging as well.

nicksellen commented:
Yeah, websockets would be better, in 2020 I think it's safe enough to use plain websockets without socket.io now. It would be relatively easy to add it on top (from the client code perspective). The server would be more work, so would leave it out for now.

nicksellen commented on modules/messages/client/components/UnreadCount.component.js:
```diff
@@ -0,0 +1,28 @@
+
+  const [count, setCount] = useState(null);
+
+  useEffect(() => watchUnreadCount(setCount), []);
```
Yes I quite liked how that pattern turned out :)

Passing `setCount` to the service is just the same as doing it more verbosely, so, a full verbose version of it would be:

```js
const [count, setCount] = useState(null);

useEffect(() => {
  const unwatch = watchUnreadCount(count => {
    setCount(count);
  })
  return unwatch;
}, []);
```

In that more verbose one we're _effectively_ still passing `setCount` to the service, just wrapped up inside the anonymous arrow function that we pass it, but it's still there.

Arguably it's _less_ clear how I have it and a bit more _magic_, so open for opinions!

nicksellen commented on modules/messages/client/components/UnreadCount.component.js:
```diff
@@ -0,0 +1,28 @@
+
+  useEffect(() => watchUnreadCount(setCount), []);
+
+  if (count === null || count === 0) {
```
Yes, that was my desire, because of how my brain works, but you're right, there is no practical reason to distinguish them. I'll change it to default to 0. If we did need to distinguish them in the future it would not be hard.

mrkvon replied:
@nicksellen I'd say, stick with whatever makes more sense to you. Don't switch if you like it the way it is. These are just opinions, and you're the author.

These things don't matter to me, either way, really. I guess I just had nothing else to comment to, yesterday. :slightly_smiling_face:

nicksellen commented on modules/messages/client/config/messages.client.run.js:
```diff
@@ -0,0 +1,10 @@
+  enableVisibilityWatching();
+  enableFaviconUpdater();
+  // Without setImmediate the stuff in angular-compat cannot load as it's too early
+  setImmediate(enableUnreadMessageCountPolling);
```
Ah, I shouldn't use it, I somehow thought it was standard javascript, but it's more something from nodejs world --> https://developer.mozilla.org/en-US/docs/Web/API/Window/setImmediate

The purpose is to delay the enabling of the polling just a little bit, as it uses some angular stuff via the angular-compat import, which relies on angular having finished it's initializing process, and this part of the code is just a little too early for that to be ready.

It's something that would go away after fully moving away from angular.

I'll fix it to use `setTimeout` instead

simison replied:
You might need [Angular `$timeout`](https://docs.angularjs.org/api/ng/service/$timeout) to ensure Angular has finished initializing.

Aside, window/global differences between environments reminded me of `globalThis` :-)  https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/globalThis

nicksellen replied:
Yup, that might be better, given we can load angular stuff here... I'll try it out.

nicksellen replied:
Also works, so I'll keep it with `$timeout`.

mrkvon commented:
So how does this work? My understanding is as follows: There is a service that starts running at the beginning and as long as a user is signed in, it regularly asks api about new messages. When the data is fetched, it can execute callbacks provided to it via the watch method.
We execute two such callbacks. In one case we update the favicon (no react involved), in other case we update the state of UnreadCount component via useEffect hook.
There is also a piece that changes frequency of fetching when we look at the app or not. That piece is an optimalization; not crucial for this to work.
How precise is this? What am i missing?

How would this system work if we had a purely React app? Would it be similar, or shaped differently?

My understanding is as follows:
In React + redux world, there would be an effect (e.g. redux-thunk) that would fetch the data regularly (or via socket), update redux state, and the state would get displayed in the UnreadCount component; the tooltip could be changed perhaps with `react-helmet`. In React without redux it may be similar, fetching the data in a ancestor component that react-helmet and UnreadCount would have in common.

I'm not asking for any changes in this comment. I just try to understand how this system works and how it fits the bigger picture.
I find it illuminating to see that it's possible to bypass the situation of: React leaves on top of the angular tree. React leaves which don't have react in common.
It must have been challenging to figure this out.

nicksellen commented:
> How precise is this? What am i missing?

Yes, that's correct! Nothing missing.

> How would this system work if we had a purely React app? Would it be similar, or shaped differently?

I don't think I can say a definitive version, but it wouldn't be so dissimilar. Assuming we left out the websockets for now still, instead of the service managing it's own subscribers it could trigger an update in the redux store. And anything dependent on that data could then receive updates via the store.

Another way to keep the services and the implementation of the store (redux) decoupled, would be to keep the services as they are, and have a small adapter that subscribes to the services, then updates the store (and then still make the dependent things get their state from the store as above).

> the tooltip could be changed perhaps with react-helmet

(assuming you meant to write the _favicon_ not _tooltip_) - it doesn't seem to have anything specific for _updating_ values dynamically, but I guess the root component (the highest level one that uses a `<Helmet>` component) could set it's values via some redux state. But that's not really making much use of `Helmet`, using it to set the page title on other routes seems like it might be useful though.

As setting the favicon based on dependent state is pretty simple, I would probably favour a small adjustment to what is in this branch, but just to get it's state from redux.

(also, in all the place I say _redux_, it could be any other centralized state management approach)

> I find it illuminating to see that it's possible to bypass the situation of: React leaves on top of the angular tree. React leaves which don't have react in common. It must have been challenging to figure this out.

We are not constrained by our frameworks :) React itself works quite nicely as one piece of a toolkit for frontend development, you can use it with other structures and tools to be something that manages the whole state and page, but equally it's quite happy with a much more minimal role of rendering selected parts on a page.

nicksellen commented:
Ok, rebased, switched to `$timeout`, answered questions, and fixed a bug I hadn't noticed (https://github.com/Trustroots/trustroots/pull/1370/commits/adeb3a9a6cf470d6c1091d97666d2dfe1eb03957).

nicksellen commented:
> I must admit that much of this still evades my grasp.

Actually, I think this means we should soon try and get some of this global state stuff into a standard solution (e.g. redux), as then it might be easier to understand as it would follow some conventions that exist elsewhere. Maybe it would be helpful to do some state management exploration, a bit like how it was for [testing approaches](https://github.com/Trustroots/trustroots/pull/1147#issuecomment-565741145)?

Let's see if @simison wants to add anything to that, before merging.

simison commented:
@nicksellen @mrkvon sorry I couldn't find time to look into this over the weekend — feel free to merge and don't wait for me! :-)

I do echo @mrkvon 's sentiments that it might be a little hard to grok, so making sure there's enough documentation for anyone to understand this bit of code would be important.

Also agreed about timing potentially being ripe for thinking about data-layer. :-)

nicksellen commented:
Thanks for input, issue for data layer discussion --> https://github.com/Trustroots/trustroots/issues/1416
